build_tools/circle/push_doc.sh

```diff
@@ -3,33 +3,74 @@
 # circle.yml. See https://circleci.com/docs/ for more details.
 # The behavior of the script is controlled by environment variable defined
 # in the circle.yml in the .circleci folder of the project.
+# Test that the vars have been set
+if [ -z ${CIRCLE_BRANCH+x} ]; then echo "CIRCLE_BRANCH is unset"; fi
+if [ -z ${CIRCLE_SHA1+x} ]; then echo "CIRCLE_SHA1 is unset"; fi
+if [ -z ${DOC_REPO+x} ]; then echo "DOC_REPO is unset"; fi
+if [ -z ${ORGANIZATION+x} ]; then echo "ORGANIZATION is unset"; fi
+if [ -z ${HOME+x} ]; then echo "HOME is unset"; fi
+if [ -z ${EMAIL+x} ]; then echo "EMAIL is unset"; fi
+if [ -z ${USERNAME+x} ]; then echo "USERNAME is unset"; fi
 
-if [ "$CIRCLE_BRANCH" = "master" ]
+if [[ "$CIRCLE_BRANCH" =~ ^master$|^[0-9]+\.[0-9]+\.X$ ]]
 then
-    dir=dev
+  echo Pushing $CIRCLE_BRANCH to gh-pages branch...
+
+  if [ "$CIRCLE_BRANCH" = "master" ]
+  then
+      dir=dev
+  else
+      # Strip off .X
+      dir="${CIRCLE_BRANCH::-2}"
+  fi
+
+  MSG="Pushing the docs to $dir/ for branch: $CIRCLE_BRANCH, commit $CIRCLE_SHA1"
+
+  cd $HOME
+  if [ ! -d $DOC_REPO ];
+  then git clone "git@github.com:"$ORGANIZATION"/"$DOC_REPO".git";
+  fi
+  cd $DOC_REPO
+  git config core.sparsecheckout true
+  echo $dir > .git/info/sparse-checkout
+  git fetch origin gh-pages
+  git checkout gh-pages
+  git reset --hard origin/gh-pages
+  git rm -rf $dir/ && rm -rf $dir/
+  cp -R $HOME/project/doc/_build/html $dir
+  git config --global user.email $EMAIL
+  git config --global user.name $USERNAME
+  git config --global push.default matching
+  git add -f $dir/
+  git commit -m "$MSG" $dir
+
+  echo $MSG
+  git push origin gh-pages
+
+  echo "Done"
+
 else
-    # Strip off .X
-    dir="${CIRCLE_BRANCH::-2}"
-fi
+  # We run the git stuff (except the push) to check that it would've worked
+  echo "Testing git clone and commit (but not push) for $CIRCLE_BRANCH"
+  dir=dev
 
-MSG="Pushing the docs to $dir/ for branch: $CIRCLE_BRANCH, commit $CIRCLE_SHA1"
+  cd $HOME
+  if [ ! -d $DOC_REPO ];
+  then git clone "git@github.com:"$ORGANIZATION"/"$DOC_REPO".git";
+  fi
+  cd $DOC_REPO
+  git config core.sparsecheckout true
+  echo $dir > .git/info/sparse-checkout
+  git fetch origin gh-pages
+  git checkout gh-pages
+  git reset --hard origin/gh-pages
+  git rm -rf $dir/ && rm -rf $dir/
+  cp -R $HOME/project/doc/_build/html $dir
+  git config --global user.email $EMAIL
+  git config --global user.name $USERNAME
+  git config --global push.default matching
+  git add -f $dir/
+  git commit -m "Commit test" $dir
 
```

Comment:
It's a good idea to test this stuff if not in master, but I would not copy the script to make sure the same script is used in both cases. Maybe set a variable in the check for the branch name, and use it later to decide if the changes shall be pushed. 

Other than that, it looks good!

Reply:
OK, will do